File: app/src/main/java/ch/bailu/aat/gpx/tools/Inverser.java
package ch.bailu.aat.gpx.tools;

import ch.bailu.aat.gpx.AltitudeDelta;
import ch.bailu.aat.gpx.AutoPause;
import ch.bailu.aat.gpx.GpxList;
import ch.bailu.aat.gpx.GpxListArray;
import ch.bailu.aat.gpx.GpxPoint;
import ch.bailu.aat.gpx.GpxPointFirstNode;
import ch.bailu.aat.gpx.GpxPointNode;
import ch.bailu.aat.gpx.MaxSpeed;

public class Inverser {


    private final GpxList newList;
    private final GpxListArray listInverse;


    public Inverser(GpxList track) {
        newList = new GpxList(track.getDelta().getType(),
                MaxSpeed.NULL,  AutoPause.NULL, AltitudeDelta.NULL);


        GpxListArray list = new GpxListArray(track);
        listInverse = new GpxListArray(track);

        if (list.size() > 0) {

            int indexInverse = list.size() -1;
            int index = 0;

            while (indexInverse >=0) {
                listInverse.setIndex(indexInverse);
                list.setIndex(index);

                final GpxPointNode point = list.get();
                final GpxPointNode pointInverse = listInverse.get();

                final GpxPoint pointNew = new GpxPoint(pointInverse, pointInverse.getAltitude(),
                        point.getTimeStamp());


                if (isLastInSegment(pointInverse))
                    newList.appendToNewSegment(pointNew, pointInverse.getAttributes());
                else
                    newList.appendToCurrentSegment(pointNew, pointInverse.getAttributes());


                index++;
                indexInverse--;
            }
        }

    }


    private boolean isLastInSegment(GpxPointNode point) {
        return point.getNext() == null || point.getNext() instanceof GpxPointFirstNode;
    }

    public GpxList getNewList() {
        return newList;
    }
}

File: aat-lib/src/main/java/ch/bailu/aat_lib/gpx/tools/Inverser.kt
package ch.bailu.aat_lib.gpx.tools

import ch.bailu.aat_lib.gpx.GpxList
import ch.bailu.aat_lib.gpx.GpxListArray
import ch.bailu.aat_lib.gpx.GpxPoint
import ch.bailu.aat_lib.gpx.GpxPointFirstNode
import ch.bailu.aat_lib.gpx.GpxPointNode
import ch.bailu.aat_lib.gpx.attributes.GpxListAttributes

class Inverser(track: GpxList) {
    val newList: GpxList = GpxList(
        track.getDelta().getType(),
        GpxListAttributes.NULL
    )

    init {
        val list = GpxListArray(track)
        val listInverse = GpxListArray(track)
        if (list.size() > 0) {
            var indexInverse = list.size() - 1
            var index = 0
            while (indexInverse >= 0) {
                listInverse.index = indexInverse
                list.index = index
                val point = list.get()
                val pointInverse = listInverse.get()
                val pointNew = GpxPoint(
                    pointInverse, pointInverse.getAltitude().toFloat(),
                    point.getTimeStamp()
                )
                if (isLastInSegment(pointInverse)) newList.appendToNewSegment(
                    pointNew,
                    pointInverse.getAttributes()
                ) else newList.appendToCurrentSegment(pointNew, pointInverse.getAttributes())
                index++
                indexInverse--
            }
        }
    }

    private fun isLastInSegment(point: GpxPointNode): Boolean {
        return point.next == null || point.next is GpxPointFirstNode
    }
}
